bpf: fix drop metric in handle_netdev()

julianwiedmann opened the pull request:
If the host firewall sends a drop-notification for an unknown L2 protocol in the from-netdev path, me thinks we need to use METRIC_INGRESS.

Subsequent code in do_netdev() agrees.

Fixes: 88bf29180099 ("bpf: Enforce host policies for IPv4")
Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1022,7 +1022,7 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
 
 		return send_drop_notify(ctx, SECLABEL, WORLD_ID, 0, ret,
-					CTX_ACT_DROP, METRIC_EGRESS);
+					CTX_ACT_DROP, METRIC_INGRESS);
```
Is actually depends whether this is called from the `to-netdev` or `to-host` section. We'll need to check `from_host` variable to know.

julianwiedmann replied:
> Is actually depends whether this is called from the `to-netdev` or `to-host` section. We'll need to check `from_host` variable to know.

Oh right. If `from-host` wants `METRIC_EGRESS` then we have it wrong in quite a few more places across `bpf_host` :smile:.

pchaigno reviewed (changes requested):
Nice catch!

julianwiedmann committed:
bpf: fix metric direction in from-host encrypt path

If a packet gets dropped while redirecting after encryption (or while
rebuilding it in preparation for this), it's on the egress part of its
journey. Set the direction in the drop notification accordingly.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: fix metric direction for mixed from-host/from-netdev paths

Set the correct direction when from-host packets are dropped early on
their journey by eg. the host firewall.

Note that this doesn't solve *all* similar cases.
In tail_handle_ipv*(), we need to report drops that occur on both parts of
a from-host packet's journey - it should be METRIC_EGRESS if dropped by the
host firewall, but METRIC_INGRESS if eg. an error occurred later in
ipv*_local_delivery().

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: untangle from-host and from-netdev paths in do_netdev()

In preparation for a subsequent change, reorder the code to create a simple
if-else flow.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -76,7 +76,7 @@ static __always_inline int rewrite_dmac_to_host(struct __ctx_buff *ctx,
 	/* Rewrite to destination MAC of cilium_net (remote peer) */
 	if (eth_store_daddr(ctx, (__u8 *) &cilium_net_mac.addr, 0) < 0)
 		return send_drop_notify_error(ctx, src_identity, DROP_WRITE_ERROR,
-					      CTX_ACT_OK, METRIC_INGRESS);
```
I'm not against that change but just want to note things are a bit more subtle than they may appear:
Ingress or egress is always with respect to something. In the case of host policy enforcement, it's easy: we're talking of egress/ingress out of/into the host-network namespace. Same for pod policies with pod-network namespaces. But for a packet traveling from pod-to-pod through `cilium_host` with host firewall disabled, should it be considered ingress or egress? :grin:

julianwiedmann replied:
Egh. Yes, I see what you mean. Feels like we're right in the middle of the packet's journey here. And given that it's an unexpected error, probably not worth changing.

julianwiedmann replied:
> In the case of host policy enforcement, it's easy: we're talking of egress/ingress out of/into the host-network namespace.

This just made me have another look at the `from-host` path. If `ipv4_host_policy_egress()` returns an error right now, `tail_handle_ipv4()` reports the drop with `METRIC_INGRESS`. While we would presumably prefer `METRIC_EGRESS`, to match the policy verdict.

Feels way too entangled :/.

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -281,7 +281,8 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host)
 			return CTX_ACT_OK;
 
 		return ipv6_local_delivery(ctx, l3_off, secctx, ep,
-					   METRIC_INGRESS, from_host);
```
Given we're talking about delivery to the local pod here, I would expect it to always be "ingress".
